refactor(bb): thread chunk for better work distribution

johnathan79717 opened the pull request:
### Changes
- Added `ThreadChunk` structure for efficient work distribution across threads
- Implemented `add_scaled_chunk` method for chunked polynomial operations  
- Modified `add_scaled` to use chunked processing for better memory access patterns
- Fixed `bc` command not found errors in CPU scaling benchmark script (replaced with `awk`)

## Benchmark Results

### CPU Scaling Performance

**After optimization (this branch):**
```
CPUs     Time (ms)       Speedup      Efficiency  
──── ────────── ─────── ──────────
1        5840.54         1.00x        100.0%      
2        2968.66         1.97x        98.5%       
4        1530.68         3.82x        95.5%       
8        812.71          7.19x        89.9%       
16       706.82          8.26x        51.6%
```

**Before optimization:**
```
CPUs     Time (ms)       Speedup      Efficiency  
──── ────────── ─────── ──────────
1        5840.56         1.00x        100.0%      
2        2970.08         1.97x        98.5%       
4        1533.58         3.81x        95.2%       
8        813.08          7.18x        89.8%       
16       706.92          8.26x        51.6% 
```



### Performance Impact
- Negligible

johnathan79717 committed:
wip(bb): chunk

johnathan79717 committed:
add_scaled

johnathan79717 committed:
fix: Replace bc with awk in CPU scaling benchmark script

The script was failing with 'bc: command not found' errors. Replaced all bc
usage with awk for floating-point comparisons to ensure compatibility on
systems without bc installed.

johnathan79717 commented:
Will fix the failed test first

johnathan79717 committed:
fix: Correct polynomial indexing in add_scaled_chunk

The chunked polynomial operation was using incorrect indices, causing
assertion failures when accessing polynomials with non-zero start indices.
Fixed by adjusting the loop to use actual polynomial indices.

johnathan79717 committed:
fix: Revert chunked folding in protogalaxy to fix virtual size expansion

The chunked implementation couldn't properly handle the polynomial swap
that occurs during virtual size expansion. Reverted to the original
two-loop approach which correctly handles polynomials with different
memory layouts.

johnathan79717 commented:
tests fixed

ludamad reviewed (approved):
LGTM. Shame it wasn't a bigger gain, but the code reads well after :)

johnathan79717 committed:
chore: Add cleanup of remote benchmark files to prevent stale data

The script now removes temporary benchmark JSON files from the remote
instance before and after each benchmark run to prevent accidentally
using cached results from failed runs.

johnathan79717 commented:
I'm rerunning the bench numbers after my last fix.

johnathan79717 commented:
I updated the numbers. They look almost identical now. 😅

ludamad commented:
Thanks Sergei for pointing out that this was tested on an overflow case, so we weren't saving with the per-poly threading. 
tested on ecdsar1+transfer_0_recursions+sponsored_fpc
before
```
      └─ ProtogalaxyProver_::update_target_sum_and_fold                                    [3]  13.3 %   544.2ms    (77.74 ms x 7)     260.1ms    with no threading
        ├─ do_iterations()                                                                   [4]  54.5 %   296.6ms    (24.5 μs x 12110)  16 threads 279.952ms average 4% stddev
        ├─ spinning main thread                                                              [4]  32.3 %   175.6ms    (228.0 μs x 770)     175.6ms    with no threading
        └─ (other)                                                                           [4]  13.2 %
2nd run sanity check (on mainframe due to laziness)
      └─ ProtogalaxyProver_::update_target_sum_and_fold                                    [3]  13.5 %   474.4ms    (67.77 ms x 7)
        ├─ do_iterations()                                                                   [4]  53.9 %   255.7ms    (21.1 μs x 12110)  16 threads 239.205ms average 3% stddev
        ├─ spinning main thread                                                              [4]  32.0 %   151.7ms    (197.0 μs x 770)
        └─ (other)                                                                           [4]  14.1 %  
3rd
      └─ ProtogalaxyProver_::update_target_sum_and_fold                                    [3]  12.7 %   447.4ms    (63.91 ms x 7)
        ├─ do_iterations()                                                                   [4]  59.4 %   265.6ms    (21.9 μs x 12110)  16 threads 246.729ms average 3% stddev
        ├─ spinning main thread                                                              [4]  30.4 %   135.9ms    (176.5 μs x 770)
        └─ (other)                                                                           [4]  10.3 % 
```
after
```
      └─ ProtogalaxyProver_::update_target_sum_and_fold                                    [3]  9.5  %   364.6ms    (52.08 ms x 7)     118.4ms    with no threading
        ├─ do_iterations()                                                                   [4]  76.7 %   279.5ms    (1.25 ms x 224)  16 threads 250.718ms average 7% stddev
        └─ spinning main thread                                                              [4]  21.8 %    
2nd run sanity check (on mainframe due to laziness)
      └─ ProtogalaxyProver_::update_target_sum_and_fold                                    [3]  9.6  %   362.8ms    (51.83 ms x 7)     161.2ms    with no threading
        ├─ do_iterations()                                                                   [4]  76.4 %   277.1ms    (1.24 ms x 224)  16 threads 242.171ms average 9% stddev
        └─ spinning main thread                                                              [4]  33.8 %   122.7ms    (8.76 ms x 14)     122.7ms    with no threading
3rd run
      └─ ProtogalaxyProver_::update_target_sum_and_fold                                    [3]  9.2  %   335.0ms    (47.85 ms x 7)     96.3ms     with no threading
        ├─ do_iterations()                                                                   [4]  76.2 %   255.3ms    (1.14 ms x 224)  16 threads 230.626ms average 5% stddev
        ├─ spinning main thread                                                              [4]  17.0 %             
        └─ (other)                                                                           [4]  6.8  % 
```
